rust_src/src/lists.rs

```diff
@@ -486,3 +491,87 @@ pub fn merge(mut l1: LispObject, mut l2: LispObject, pred: LispObject) -> LispOb
         tail = item;
     }
 }
+
+/// Struct that we can use to map a sequence without storing the results.
+struct SideEffectOnly;
+
+impl<T> iter::FromIterator<T> for SideEffectOnly {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        iter.into_iter().count(); // Need to exhaust the iterator.
+        SideEffectOnly
+    }
+}
+
+/// Guts of all mapping functions.
+fn mapcar1<T: iter::FromIterator<LispObject>>(seq: LispObject, func: LispObject) -> T {
+    if seq.is_cons() {
+        seq.iter_tails_safe()
+            .map(|cons| call!(func, cons.car()))
+            .collect()
+    } else if seq.is_nil() {
+        iter::empty().collect()
+    } else if let Some(vec) = seq.as_vectorlike().and_then(|v| v.as_vector()) {
+        vec.iter().map(|item| call!(func, item)).collect()
+    } else if let Some(bvec) = seq.as_vectorlike().and_then(|v| v.as_bool_vector()) {
+        bvec.iter().map(|item| call!(func, item)).collect()
+    } else if let Some(string) = seq.as_string() {
+        string
+            .chars()
+            .map(|ch| call!(func, LispObject::from_fixnum(ch as EmacsInt)))
+            .collect()
```

Comment:
Is it possible that this is a multibyte char issue ? Can char() handle those ?

Reply:
It's not Rust `chars`, but an iterator for `LispStringRef` which handles multibyte and unibyte. But it's still possible that it is buggy, I'll have a look.

Reply:
The code is actually failing in the `is_cons` case. I added debug prints to catch it.

I tweaked mapconcat:

    +    println!("Begin {:?}", function);
    +    for i in mapped.iter() {
    +        match i.as_string() {
    +            Some(s) => println!("mapconcat(r): {} == {:?}", i.to_raw(), *s),
    +            None => println!("mapconcat(r): {}", i.to_raw()),
    +        };
    +    }
    +    println!("End");

Along with a derive(Debug) for Lisp_String to avoid sending bogus pointers to `display_string`. Here is the output for $ gdb --args ../src/remacs -Q --no-site-file --no-site-lisp -batch -L :. -l ert -l lisp/sort-tests.el

    Begin '('closure '('('n . 8 't ) '('_ ) '('string '('let '('('c '('random 52 ) ) ) '('+ '('if '('> 'c 25 ) 71 65 ) 'c ) ) ) )
    mapconcat(r): 16041668 == Lisp_String { size: 1, size_byte: -1, intervals: 0x0, data: 0xedba98 }
    mapconcat(r): 16041700 == Lisp_String { size: 16041600, size_byte: -1, intervals: 0x0, data: 0x0 }
    mapconcat(r): 16041732 == Lisp_String { size: 16041696, size_byte: -1, intervals: 0x0, data: 0x0 }
    mapconcat(r): 16041764 == Lisp_String { size: 16041728, size_byte: -1, intervals: 0x0, data: 0x0 }
    mapconcat(r): 16041796 == Lisp_String { size: 16041760, size_byte: -1, intervals: 0x0, data: 0x0 }
    mapconcat(r): 16041828 == Lisp_String { size: 1, size_byte: -1, intervals: 0x0, data: 0xedbaa8 }
    mapconcat(r): 15757748 == Lisp_String { size: 1, size_byte: -1, intervals: 0x0, data: 0xedbab8 }
    mapconcat(r): 15757716 == Lisp_String { size: 1, size_byte: -1, intervals: 0x0, data: 0xedbac8 }
    End

The data is bad straight out of the invocation by `call!` near as I can tell.

Reply:
I debugged this a little further today.

    seq.iter_tails_safe()
            .map(|cons| {
                let result = call!(func, cons.car());
                println!("Call!: {} -> {:?}", result.to_raw(), result);
                result
            })
            .collect();

The data is printed correctly in the `map`. If I switch to a loop and manually populate the vector I can see that the data goes south as the vector is being filled:

     for i in seq.iter_tails_safe() {
            let result = call!(func, i.car());
            println!("Call!: {} -> {:?}", result.to_raw(), result);
            output.push(result);
            for g in output.iter() {
                match g.as_string() {
                    Some(_) => {
                        println!("mapcar1(r): {}", g.to_raw());
                        println!("== {:?}", g);
                    }
                    None => println!("mapcar1(r): {}", g.to_raw()),
                }
            }

Now you can see that after a few iterations the pointers are off and becoming garbage. It has nothing to do with the call to `random` in the elisp test. I dummied out the function using `/dev/urandom` and it still fails.